File: constraint-substrate/c/src/holonomy.c
#include "constraint_substrate.h"

double cs_holonomy(const double* values, uint32_t count, double modulus) {
    if (count == 0) return 0.0;
    double total = 0.0;
    for (uint32_t i = 1; i < count; i++) {
        double diff = values[i] - values[i - 1];
        double wrapped = diff - modulus * round(diff / modulus);
        total += wrapped;
    }
    return total / modulus;
}

File: constraint-substrate/python/constraint_substrate/holonomy.py
"""Holonomy / winding number computation."""


def winding(values: list[float], modulus: float) -> float:
    """Compute holonomy (winding number) of a sequence of values modulo `modulus`."""
    if not values:
        return 0.0

    total = 0.0
    for i in range(1, len(values)):
        diff = values[i] - values[i - 1]
        wrapped = diff - modulus * round(diff / modulus)
        total += wrapped

    return total / modulus
